File: 백준/Gold/1520. 내리막 길/내리막 길.py
from sys import stdin
input = stdin.readline

# y, x 에서 0, 0으로 이동할수 있는 모든 경우의 수를 dp에 기록한다
def solution(y, x):    
    # 목적지 도착 여부
    if y == M-1 and x == N-1:
        dp[y][x] = 1
        return
    
    # 이미 방문한곳
    if dp[y][x] != -1:
        return
    
    # 방문 표시
    dp[y][x] = 0
    
    # 상, 하, 좌, 우
    for dy, dx in [[y-1, x], [y+1, x], [y, x-1], [y, x+1]]:
        if 0 <= dy < M and 0 <= dx < N:
            if road[y][x] > road[dy][dx]:
                solution(dy, dx)
                dp[y][x] += dp[dy][dx]

M, N = map(int, input().split())
road = [list(map(int, input().split())) for _ in range(M)]
dp = [[-1] * N for _ in range(M)]
solution(0, 0)
print(dp[0][0])

File: 백준/Gold/1520. 내리막 길/내리막 길.java
public class Main {
    public static void main(String[] args) throws Exception {
        M = read();
        N = read();

        road = new int[M][N];
        dp = new int[M][N];

        for (int y = 0; y < M; y++) {
            for (int x = 0; x < N; x++) {
                road[y][x] = read();
                dp[y][x] = -1;
            }
        }

        solution(0, 0);
        System.out.print(dp[0][0]);
    }

    static int M;
    static int N;
    static int[][] road;
    static int[][] dp;

    private static void solution(int y, int x) {
        if (y == M - 1 && x == N - 1) {
            dp[y][x] = 1;
            return;
        }

        if (dp[y][x] != -1) {
            return;
        }

        dp[y][x] = 0;
        for (int[] temp : new int[][]{{y - 1, x}, {y + 1, x}, {y, x - 1}, {y, x + 1}}) {
            int dy = temp[0];
            int dx = temp[1];

            if (0 <= dy && dy < M && 0 <= dx && dx < N) {
                if (road[y][x] > road[dy][dx]) {
                    solution(dy, dx);
                    dp[y][x] += dp[dy][dx];
                }
            }
        }
    }

    private static int read() throws Exception {
        int c, n = System.in.read() & 15;
        while ((c = System.in.read()) > 32) n = (n << 3) + (n << 1) + (c & 15);
        return n;
    }
}